website/server/libs/applePayments.js

```diff
@@ -0,0 +1,172 @@
+          amount = 5.25;
+          break;
+        case 'com.habitrpg.ios.Habitica.42gems':
+          amount = 10.5;
+          break;
+        case 'com.habitrpg.ios.Habitica.84gems':
+          amount = 21;
+          break;
+      }
+      if (amount) {
+        correctReceipt = true;
+        await payments.buyGems({ // eslint-disable-line no-await-in-loop
+          user,
+          paymentMethod: api.constants.PAYMENT_METHOD_APPLE,
+          amount,
+          headers,
+        });
+      }
+    }
+  }
+
+  if (!correctReceipt) throw new NotAuthorized(api.constants.RESPONSE_INVALID_ITEM);
+
+  return appleRes;
+};
+
+api.subscribe = async function subscribe (sku, user, receipt, headers, nextPaymentProcessing = undefined) {
+  if (!sku) throw new BadRequest(shared.i18n.t('missingSubscriptionCode'));
+  let subCode;
+  switch (sku) {
+    case 'subscription1month':
+      subCode = 'basic_earned';
+      break;
+    case 'com.habitrpg.ios.habitica.subscription.3month':
+      subCode = 'basic_3mo';
+      break;
+    case 'com.habitrpg.ios.habitica.subscription.6month':
+      subCode = 'basic_6mo';
+      break;
+    case 'com.habitrpg.ios.habitica.subscription.12month':
+      subCode = 'basic_12mo';
+      break;
+  }
+  let sub = subCode ? shared.content.subscriptionBlocks[subCode] : false;
+  if (!sub) throw new NotAuthorized(this.constants.RESPONSE_INVALID_ITEM);
+
+  await iap.setup();
+
+  let appleRes = await iap.validate(iap.APPLE, receipt);
+  let isValidated = iap.isValidated(appleRes);
+  if (!isValidated) throw new NotAuthorized(api.constants.RESPONSE_INVALID_RECEIPT);
+
+  let purchaseDataList = iap.getPurchaseData(appleRes);
+  if (purchaseDataList.length === 0) throw new NotAuthorized(api.constants.RESPONSE_NO_ITEM_PURCHASED);
+
+  let transactionId;
+
+  for (let index in purchaseDataList) {
+    let purchaseData = purchaseDataList[index];
+
+    let dateTerminated = new Date(Number(purchaseData.expirationDate));
+    if (purchaseData.productId === sku && dateTerminated > new Date()) {
+      transactionId = purchaseData.transactionId;
+      break;
+    }
+  }
+  if (transactionId) {
+    let existingUser = await User.findOne({
+      'purchased.plan.customerId': transactionId,
+    }).exec();
+    if (existingUser) throw new NotAuthorized(this.constants.RESPONSE_ALREADY_USED);
+
+    nextPaymentProcessing = nextPaymentProcessing ? nextPaymentProcessing : moment.utc().add({days: 2});
+
+    await payments.createSubscription({
+      user,
+      customerId: transactionId,
+      paymentMethod: this.constants.PAYMENT_METHOD_APPLE,
+      sub,
+      headers,
+      nextPaymentProcessing,
+      additionalData: receipt,
+    });
+  } else {
+    throw new NotAuthorized(api.constants.RESPONSE_INVALID_RECEIPT);
+  }
+};
+
+
+api.cancelSubscribe = async function cancelSubscribe (user, headers) {
+  let data = user.purchased.plan.additionalData;
+
+  if (!data) throw new NotAuthorized(shared.i18n.t('missingSubscription'));
```

Comment:
This might be irrelevant since you're talking about a new subscription payment method, but I've never seen purchased.plan.additionalData in any of the subscriptions I've ever looked at.